tokenリストから抽象構文木を作成

nafuka11 opened the pull request:
Fix #5

## やったこと

tokenリストから抽象構文木（t_node）を作成し、dotファイル（ast.dot）で表示できるようにしました。
コマンド実行はまだできません。

## 各関数の説明

### srcs/parser/parser.c

関数名 | 内容
-- | --
parse_complete_command | tokenリストを処理する関数。parse_pipeline の呼び出し結果が FALSE or 処理後tokenが残るなら FALSE
parse_pipeline | pipelineを処理する関数。parse_command の呼び出し結果が FALSE なら FALSE
parse_command | commandを処理する関数。command何もなし or parse_io_redirect の呼び出し結果が FALSE なら FALSE
parse_io_redirect | redirectを処理する関数。redirectのtokenが不正 or redirectの後がTOKENでなければFALSE

### srcs/parser/parser_utils.c

関数名 | 内容
-- | --
add_copied_token | 指定tokenをコピーし、tokenリストに追加する。
copy_token | 指定tokenをコピーしたtokenを返す。
has_token_type | nodeが指定typeか判定する関数。判定しつつtokenを進める。
is_redirect_token | token_typeがredirectか判定する関数。

### srcs/parser/node.c

関数名 | 説明
-- | --
add_parent_node | 親nodeをmallocし、node_typeと子nodeのポインタをセットする関数。
set_command_args | t_command->args にtokenをコピーしつつtokenを進める関数。
create_command_node | commandを持つnodeをmallocする関数。
del_node_list | nodeを全てfreeする関数。nodeはNULLセットしない（ダブルポインタにすると再帰できないと思うので）。

### srcs/parser/redirect.c

関数名 | 説明
-- | --
add_redirect | redirectリストに指定redirectを追加する関数。
create_redirect | redirectを新規作成する関数。
set_redirect_type | token_typeに応じたredirect_typeをセットする関数。
del_redirect_list | redirectリストを削除する関数。

### srcs/utils/error.c

関数名 | 説明
-- | --
print_unexpected_token_error | tokenに対応したエラーを表示する関数。

nafuka11 committed:
feat(exec): t_command構造体追加

t_commandは単方向連結リスト。argsは一旦t_tokenリストで持つように

nafuka11 committed:
feat(parser): 抽象構文木用の構造体t_node, t_node_typeを追加

t_nodeにt_commandを含めるように

nafuka11 committed:
feat(parser): nodeを追加する関数を作成

- add_parent_node: 指定typeの親ノードを追加
- create_command_node: コマンドノードを追加

nafuka11 committed:
feat(parser): semicolin, pipeのparseに対応

- commandはcreate_command_node()で簡易的に対応
- redirectは未対応

nafuka11 committed:
chore(parser): nodeをdot形式で標準出力に出力する関数追加

nafuka11 committed:
Merge remote-tracking branch 'origin/master' into 5-create-ast-from-token-list

nafuka11 committed:
fix(exec): 不要なt_command.nextを削除

リストではないため

nafuka11 committed:
feat(exec): t_redirect, t_redirect_typeを追加

nafuka11 committed:
Merge remote-tracking branch 'origin/master' into 5-create-ast-from-token-list

nafuka11 committed:
feat(node): node情報を標準出力ではなくファイル(ast.dot)に出力するように

nafuka11 committed:
refactor(node): node_print.cをprint_node.cにリネーム

- node_print.cの関数が増え、ファイル分割しやすいファイル名にしたいため

nafuka11 committed:
chore(exec): t_redirectのメンバにfd追加

nafuka11 committed:
feat(utils): unexpected tokenエラーを出力する関数追加

nafuka11 committed:
chore(utils): utils.hにprint_unexpected_token_error()追加

- utils.hとtoken.hが相互includeしているのでt_tokenの宣言を入れてます

nafuka11 committed:
fix(parser): norm error(break後のspace)

nafuka11 committed:
feat(parser): redirectのparseを仮実装

- redirectを上書きするだけです
- token切断してません
- '; ; ; ;' で ';'でなく'newline'tokenエラーになります

nafuka11 committed:
feat(parser): t_redirect操作関数追加(新規作成, listに追加)

create_redirect: t_redirect作成, 初期化
add_redirect: 既存redirectリストに指定redirectを追加

nafuka11 committed:
feat(parser): node出力関数でredirectを追加するように

print_node_label.c にnodeに表示する文字を処理する部分を分けました
重複処理をprint_edge()にまとめました

nafuka11 committed:
chore(parser): create_redirect, add_redirectを使用するように

nafuka11 committed:
refactor: utils.hをtoken.hでなくtoken.cでincludeするように

utils.h: 不要になったt_tokenの宣言を削除. indent修正

nafuka11 committed:
refactor(token): 不要なincludeを削除

nafuka11 committed:
Merge remote-tracking branch 'origin/master' into 5-create-ast-from-token-list

nafuka11 committed:
chore(parser): redirect判定でD_GREATを使用するように

nafuka11 committed:
feat(parser): tokenをコピーする関数追加

nafuka11 committed:
fix(parser): normエラー修正

set_redirect_type()をredirect.cへ移動
文法のparseはparser.cに、細々した関数はparser_utils.cへ移動

nafuka11 committed:
feat(parser): エラーハンドリング追加。parse系関数でparse成否を返すように

parse失敗時にunexpected tokenエラーを表示する

nafuka11 committed:
chore(node): print_node_label.cの改行修正

nafuka11 committed:
chore(Makefile): SRCS更新

nafuka11 committed:
feat(parser): t_redirectをfreeする関数追加

nafuka11 committed:
feat(parser): t_nodeをfreeする関数追加

nafuka11 committed:
feat: mainでlexer, parserを実行するように

nafuka11 committed:
fix(parser): redirect token判定にD_GREATER追加

ToYeah commented on srcs/parser/parser.c:
```diff
@@ -0,0 +1,121 @@
+	redirect = create_redirect();
+	if ((*tokens)->type == IO_NUMBER)
+	{
+		redirect->fd = ft_atoi((*tokens)->data);
```
オーバーフローすると小っちゃくなっちゃうかもしれないです
ft_atoi_limit使えば一応オーバーフロー検知できます。
本家はfdがulimit -aで見れる設定にあるopen filesを超えるとbad file descriptorっていうエラーが出るみたいです。
open filesを取得する方法はない気がするのでどこかでMAXの値を設定するとかですかね...?

nafuka11 replied:
ご指摘ありがとうございます。
INT_MAXの範囲を超えると `file descriptor out of range` のエラー表示が追加されるので、ft_atoi_limitを使って、範囲外なら、-2を入れようと思います。-1はfd未指定扱いにしてあります。
```
bash-3.2$ echo 2147483647> file
bash: 2147483647: Bad file descriptor
bash-3.2$ echo 2147483648> file
bash: file descriptor out of range: Bad file descriptor
```

`bad file descriptor` は不正なfdに書き込みなどを行った時にセットされるerrnoのメッセージのようです。
環境依存だと思うので、MAXは設定しないでいいのではないかと思います。
以下のコードで `bad file descriptor` を再現できます。
```
#include <unistd.h>
#include <stdio.h>
#include <errno.h>

int	main(void)
{
	write(2147483648, "a", 1);
	perror("minishell");
	return (0);
}
```

nafuka11 replied:
93b9e2f でオーバーフローの対処を行いました。

nafuka11 committed:
fix(node): node->commandをNULL初期化

nafuka11 committed:
fix(node): node->typeがCOMMANDの時のみcommandをfreeするように

nafuka11 committed:
fix(parser): leak修正

nafuka11 commented:
`echo |` や `echo ;` などのケースでsegfaultになるのを修正しました。

nafuka11 committed:
fix(parser): redirectのfdがINT_MAXを超える場合-2を格納

INT_MAXを超えた際file descriptor out of rangeを表示させるため

nafuka11 committed:
refactor: fdの-1, -2をdefineで定義するように

ast.dotではfd範囲外はそのまま表示する(-2>file)

ToYeah commented on srcs/parser/parser.c:
```diff
@@ -0,0 +1,126 @@
+	return (TRUE);
+}
+
+static t_bool	parse_command(t_node **node, t_token **tokens)
```
26行でnormエラーが出るみたいです。

nafuka11 replied:
ありがとうございます！
7ae6203 で修正しました。

ToYeah commented on srcs/parser/parser.c:
```diff
@@ -0,0 +1,126 @@
+		if (parse_pipeline(nodes, tokens) == FALSE)
+			return (FALSE);
+	}
+	while (*tokens)
+	{
+		if (has_token_type(tokens, CHAR_SEMICOLON))
+		{
+			if (parse_pipeline(&child, tokens) == FALSE)
+				return (FALSE);
+			*nodes = add_parent_node(NODE_SEMICOLON, *nodes, child);
+		}
+		else
```
`echo a`のようなセミコロンで終わるコマンドが来ると右側の子を探して見つからないからエラー吐いちゃうみたいです！
bash本家は`echo a`が実行されます

nafuka11 replied:
4236daf で右側の子を探さないように修正しました！

nafuka11 committed:
fix(parser): redirect=-2を定数に修正

nafuka11 committed:
fix(parser): normエラー修正(parse_commandの行数削減)

nafuka11 committed:
fix(parser): コマンド;でsyntax errorになるのを修正

parse_command: 引数, redirect共にない時エラー扱いにするように
parse_complete_command: ;の後にtokenがある時だけparseするように
